### PR TITLE
[release/7.0] Add faster DAC EnumMemoryRegion option with less memory usage (#74300)

### PR DESCRIPTION
**Pull request by mikem8361**
# Customer Impact

Our partner team reports that on Linux, createdump can take up to 150mb of memory usage in a Linux container environment and may cause the container to OOM. 

Issue: https://github.com/dotnet/runtime/issues/72148

# Details

Instead of drilling down into all the individual MT/MD/EEClass, etc. data structures, add the LoaderAllocator/LoaderHeaps regions directly.

Add new CLRDATA_ENUM_MEM_HEAP2 flag for the fast path.

To reduce risk of incomplete core dumps this is enabled by the COMPlus_EnableFastHeapDumps env var. This env var is only looked at by the Linux/MacOS createdump. It is currently ignored on Windows. The new HEAP2 flag is works when passed to the EnumMemoryRegions API on Windows but createdump can't set it because MiniDumpWriteDump in dbghelp.dll loads/calls the DAC API.

Fix MacOS dlopen error message (minor createdump logging change for VS4Mac).

# Testing

Local and SOS testing.

# Risk

Low risk. Affects only the DAC under an opt-in env var.

### CODE REVIEW

**Review by leculver — APPROVED** [review]
I assume performance tracing showed that this was what was causing the slowdown.  Do we have a performance test before/after?  How much time does this save overall?

The best way to get test coverage of this is to simply run all SOS commands before/after this change and make sure nothing breaks.  Since this has a lot to do with loader heaps, commands like `!eeheap`, `!dumpdomain`, `!dumpmodule`, `!dumpclass`, `!name2ee`, and the basics of `!threads`, `!dumpheap`, and `!verifyheap` should cover everything...IIRC.

**Comment by leculver** [comment]
With respect to https://github.com/dotnet/runtime/issues/72148, does using this new flag reduce overall memory usage?  In addition to my question on how much faster this is, is there a comparison of peak to regular memory usage?

Sorry to hoist additional perf questions here, but I wanted to double-check the memory and wall-time improvements before calling related issues fixed when we report back to partner teams.

**Comment by mikem8361** [comment]
In my memory usage profiling (via Valgrind) against my "BigApp" test app with 100's of assemblies, 30 or so types and methods in each assembly (something like a large real world app) went from memory usage of 300MB to almost no memory.  Performance wise it went from minutes to generate the dump to secs.

This change saves on both createdump/DAC memory usage and wall clock time to generate the dump.

File: src/coreclr/inc/clrdata.idl [hunk 1]
```diff
@@ -308,10 +308,12 @@ interface ICLRDataLoggingCallback : IUnknown
...
-    CLRDATA_ENUM_MEM_MINI = CLRDATA_ENUM_MEM_DEFAULT,       // generating skinny mini-dump
-    CLRDATA_ENUM_MEM_HEAP = 0x1,        // generating heap dump
-    CLRDATA_ENUM_MEM_TRIAGE = 0x2,      // generating triage mini-dump
-
+    CLRDATA_ENUM_MEM_MINI = CLRDATA_ENUM_MEM_DEFAULT,   // generating skinny mini-dump
+    CLRDATA_ENUM_MEM_HEAP = 0x1,                        // generating heap dump
+    CLRDATA_ENUM_MEM_TRIAGE = 0x2,                      // generating triage mini-dump
+    /* Generate heap dumps faster with less memory usage than CLRDATA_ENUM_MEM_HEAP by adding
+       the loader heaps instead of traversing all the individual runtime data structures. */
+    CLRDATA_ENUM_MEM_HEAP2 = 0x3,
```
[the remark this answers is not in the document]
Comment (mikem8361) [reply src/coreclr/inc/clrdata.idl:316]: I know the enum name could be better but I'd rather not change it at this stage. It is used in almost every file changed.  These changes are already in main (8.0) and this is "backport" to 7.0.  I've already run the SOS tests (which cover all the commands you mentioned) on this version.

**Review by jeffschwMSFT — APPROVED** [review]
approved

**Comment by carlossanlop** [comment]
Approved and signed off.
CI is green.
Ready to merge. :shipit:

**Pull request merged**